DSharpPlus.Commands/CommandsExtension.cs

```diff
@@ -0,0 +1,233 @@
+
+    /// <inheritdoc cref="CommandsConfiguration.UseDefaultCommandErrorHandler"/>
+    public bool UseDefaultCommandErrorHandler { get; init; }
+
+    public CommandExecutor CommandExecutor { get; init; } = new();
+
+    /// <summary>
+    /// The registered commands that the users can execute.
+    /// </summary>
+    public IReadOnlyDictionary<string, Command> Commands { get; private set; } = new Dictionary<string, Command>();
+    private readonly List<CommandBuilder> _commandBuilders = [];
+
+    public IReadOnlyDictionary<Type, ICommandProcessor> Processors { get; private set; } = new Dictionary<Type, ICommandProcessor>();
+    private readonly Dictionary<Type, ICommandProcessor> _processors = [];
+
+    /// <summary>
+    /// Executed everytime a command is finished executing.
+    /// </summary>
+    public event AsyncEventHandler<CommandsExtension, CommandExecutedEventArgs> CommandExecuted { add => this._commandExecuted.Register(value); remove => this._commandExecuted.Unregister(value); }
+    internal readonly AsyncEvent<CommandsExtension, CommandExecutedEventArgs> _commandExecuted = new("COMMANDS_COMMAND_EXECUTED", EverythingWentWrongErrorHandler);
+
+    /// <summary>
+    /// Executed everytime a command has errored.
+    /// </summary>
+    public event AsyncEventHandler<CommandsExtension, CommandErroredEventArgs> CommandErrored
+    {
+        add
+        {
+            if (this.UseDefaultCommandErrorHandler)
+            {
+                this._commandErrored.Unregister(DefaultCommandErrorHandlerAsync);
+            }
+
+            this._commandErrored.Register(value);
+        }
+        remove => this._commandErrored.Unregister(value);
+    }
+    internal readonly AsyncEvent<CommandsExtension, CommandErroredEventArgs> _commandErrored = new("COMMANDS_COMMAND_ERRORED", EverythingWentWrongErrorHandler);
+
+    /// <summary>
+    /// Used to log messages from this extension.
+    /// </summary>
+    private readonly ILogger<CommandsExtension> _logger;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="CommandsExtension"/> class.
+    /// </summary>
+    /// <param name="configuration">The configuration to use.</param>
+    internal CommandsExtension(CommandsConfiguration configuration)
+    {
+        ArgumentNullException.ThrowIfNull(configuration);
+
+        this.ServiceProvider = configuration.ServiceProvider;
+        this.DebugGuildId = configuration.DebugGuildId;
+        this.UseDefaultCommandErrorHandler = configuration.UseDefaultCommandErrorHandler;
+        if (this.UseDefaultCommandErrorHandler)
+        {
+            // We don't do `this.CommandErrored += DefaultCommandErrorHandlerAsync` because that'll immediately remove the event handler.
+            this._commandErrored.Register(DefaultCommandErrorHandlerAsync);
+        }
+
+        // Attempt to get the user defined logging, otherwise setup a null logger since the D#+ Default Logger is internal.
```

Comment:
I think we should open internals when merging

Reply:
we should generally revamp that logger